Dds 643 oit oauth

dmlond opened the pull request:
implemented ability to create an openid authentication service object to interact with duke oit openid OAUTH implementation

dmlond committed:
AuthenticationService is now the base for Single Table Inheritance

dmlond committed:
refactored original AuthenticationService (non STI) into separate
DukeAuthenticationService and OitAuthenticationService which are

each Single Table Inheritance models based on the AuthenticationService
base into authentication_service_shared_example 'an authentication service'

refactored authentication_services factory into separate duke_authentication_service
and oit_authentication_service factories

refactored test for shared AuthenticationService expectations into
authentication_service_

dmlond committed:
UserSpec expects User to have a current_user_authentication_service attr
User meets spec

dmlond committed:
any authentication_service should implement a get_user_for_access_token method

dmlond committed:
DukeAuthenticationServiceSpec expects DukeAuthenticationService to implement get_user_for_access_token
- returns a User with a user.current_user_authentication_service set
- user returned is a new user, with a newly built user_authentication_service if the user has not used the DukeAuthenticationService before
- raises a JWT::VerificationError if the token is not signed with the Rails secret

dmlond committed:
DukeAuthenticationServiceSpec expects DukeAuthenticationService to implement
get_user_for_access_token:
  - should return a user with user.current_user_authentication_service set
  - user and user.current_user.authenticaton_service should be persisted if the user has logged
    in with the authentication service before
  - user and user.current_user_authentication_service should not be persisted if the user has never
    logged in
  - user should be persisted, but user.current_user_authentication_service should not be persisted
    if the user already exists based on their username, but they have never logged in with the
    authentication service before
  - should raise a JWT::VerificationError if the token is not signed with the rails secret

dmlond committed:
DukeAuthenticationServiceSpec expects DukeAuthenticationService to raise
a standard InvalidAccessTokenException instead of the JWT::VerificationError
that is only valid in the DukeAuthenticationService

dmlond committed:
pulling let variables out of get_user_for_access_token tests to prepare to
share these tests in the shared_example for 'an authentication service'

dmlond committed:
standardized get_user_for_access_token expectations and moved them to the
authentication_service_shared_examples 'an authentication service' tests

dmlond committed:
added client_id and client_secret columns to the authentication_service table
create an openid_authentication_services factory

dmlond committed:
DukeAuthenticationService does not need to set the last_login_at in the user model

Renamed OitAuthenticationService to OpenidAuthenticationService

AuthenticationService shared example an authentication service get_user_for_access_token
for existing user who has not already authenticated to the service tests that the
returned user id equals existing_first_authenticating_user.id

OpenidAuthenticationServiceSpec expects OpenidAuthenticationService to behave like
an authentication service

expects OPENID ENV set in docker-compose.dev.yml rspec service to be set

dmlond committed:
does not need oauth.env in all of the server based services

dmlond committed:
uses duke_authentication_service instead of authentication_service

dmlond committed:
uses duke_authentication_service instead of authentication_service

dmlond committed:
added is_default boolean column to authentication_services table

authentication_service shared example expects that only one AuthenticationService,
of any type, can be default in the table.
AuthenticationService validates uniqueness of is_default if the newly created object
is_default is true

added default trait to duke_authentication_services and openid_authentication_services factories

dmlond committed:
duke openid /userinfo request now returns the uid and the duke unique id so the second call to /introspect is no longer needed

Moved the before block that sets up stubbed requests to the openid /userinfo into a shared_context 'mocked openid request'

OpenidAuthenticationServiceSpec no longer expects calls to /introspec and no longer sets introspect hashes for the 'mocked openid request' context

dmlond committed:
'mocked openid request' context refactored to more reusable
'mocked openid request to', build_openid_authentication_service

dmlond committed:
added some code comments

dmlond committed:
added InvalidAuthenticationServiceIDException to handle the case when the provided service_id is not registered
authentication_service_shared_examples:
  added shared_examples an authentication request endpoint:
  - tests a request to authentication using a given authentication_service
  - ensures that proper exceptions are handled for invalid tokens, and unregistered service_ids
  - ensures that first_time users are created with a user_authentication_service
  - ensures that existing users that have not logged in with the given authentication_service create a user_authentication_service
    with the username as the user_authentication_service uid
  - tests request for identified authentication_service when the specified authentication_service is not a default service provider
UsersApiSpec
  tests requests to authenticaton with the duke_authentication_service and openid_authentication_service, both as default and specified

UsersApi
  meets all requirements specified in UsersApiSpec via 'an authentication request endpoint' shared examples

DDS::V1::Base
  adds get_auth_service helper to find get either the default authentication_service, or an identifed authentication_service

dmlond committed:
fixed duplicate, but conflicting rescue handlers for InvalidAuthenticationServiceIDException and
InvalidAccessTokenException

dmlond committed:
Merge remote-tracking branch 'official/develop' into DDS-643_OIT_OAUTH

dmlond committed:
added from_auth_env context to mock the way ENV variables are used to create and destroy these in the rake tasks

dmlond committed:
auth_service_rake_spec expects auth_service.rake to implement:
- auth_service:transfer_default
    transfers default authentication_service from ENV[FROM_AUTH_SERVICE_ID] to ENV[TO_AUTH_SERVICE_ID]
- auth_service:duke:create
    creates a DukeAuthenticationService from ENV. raises an exception if ENV[AUTH_SERVICE_IS_DEFAULT] is true
    and there is already a default authentication_service
- auth_service:duke:destroy
   destroys the DukeAuthenticationService identified by ENV[AUTH_SERVICE_SERVICE_ID]
   warns if it is the default authentication_service
- auth_service:openid:create
    creates a OpenidAuthenticationService from ENV. raises an exception if ENV[AUTH_SERVICE_IS_DEFAULT] is true
    and there is already a default authentication_service
- auth_service:openid:destroy
   destroys the OpenidAuthenticationService identified by ENV[AUTH_SERVICE_SERVICE_ID]
   warns if it is the default authentication_service

auth_service.rake meets expectations

dmlond committed:
added validations to ensure a client_id and client_secret are present when creating an OpenidAuthenticationService
ensure that calls to create AuthenticationServices in the rake task fail if they are not valid

dmlond committed:
updated AUTH_SERVICE_ID to AUTH_SERVICE_SERVICE_ID to make it easier to parse
AUTH_SERVICE_{FOO} to record.send(foo)

DDS-643 #resolve

dmlond committed:
Merge remote-tracking branch 'official/develop' into DDS-643_OIT_OAUTH

dmlond committed:
new ENV needed for openid_authentication_service tests

dmann commented on app/api/dds/v1/users_api.rb:
```diff
@@ -13,59 +13,30 @@ class UsersAPI < Grape::API
-              suggestion: 'authenticaton service not recognized'
-            }, 401)
-          end
+        raise InvalidAccessTokenException.new unless token_info_params[:access_token]
```
I'm curious if this is needed. My instinct is that an exception like this would be thrown by the auth service model when get_user_for_access_token in called. Will the tests continue to pass if line 28 is removed?

dmlond replied:
refactored into the model

dmann commented on app/models/duke_authentication_service.rb:
```diff
@@ -0,0 +1,34 @@
+  def get_user_for_access_token(encoded_access_token)
+    begin
+      access_token = JWT.decode(encoded_access_token, Rails.application.secrets.secret_key_base)[0]
+      user_authentication_service = user_authentication_services.where(uid: access_token['uid']).first
```
Does ActiveRecord have a find_by_uid that could be used instead of the where(uid: foo).first?

dmlond replied:
yes

dmann commented on app/models/duke_authentication_service.rb:
```diff
@@ -0,0 +1,34 @@
+        user.current_user_authenticaiton_service = user_authentication_service
+        user
+      else
+        user = User.where(username: access_token['uid']).take
```
Similar to above, is there a User.find_by_username method?

dmlond replied:
yes

dmann commented on lib/tasks/auth_service.rake:
```diff
@@ -1,20 +1,85 @@
+    raise '#{from_auth_service.service_id} is not default' unless from_auth_service.is_default?
+
+    to_auth_service = AuthenticationService.where(service_id: ENV['TO_AUTH_SERVICE_ID']).take!
+    from_auth_service.update(is_default: false)
```
This is probably pretty low risk, but these updates should probably be in a transaction, with a failure rolling it all back and spitting out an error.

dmlond replied:
done

dmann commented on app/models/openid_authentication_service.rb:
```diff
@@ -0,0 +1,42 @@
+
+  def get_user_for_access_token(token)
+    user_info = get_userinfo(token)
+    uid = user_info['dukeNetID']
```
Lets store these user_info keys in the database!

dmlond committed:
Merge branch 'develop' into DDS-643_OIT_OAUTH

dmann commented on spec/lib/tasks/auth_service_rake_spec.rb:
```diff
@@ -0,0 +1,154 @@
+require 'rails_helper'
+
```
Since silence_stream is used later on in this file, add the following here for Rails 5 compatibility:

```
if Rails.version > '5.0.0'              
  require 'active_support/testing/stream'  
  include ActiveSupport::Testing::Stream
end
```

dmlond replied:
done

dmann commented on spec/support/authentication_service_shared_examples.rb:
```diff
@@ -0,0 +1,452 @@
+  #   existing_user_access_token
+  #   invalid_access_token
+
+  it_behaves_like 'a GET request' do
```
I envisioned the 'a GET request' shared example as a top level call in the main request spec files that wraps calls to more functional shared examples. I really should have documented it better.  :(

dmann commented on spec/support/authentication_service_shared_examples.rb:
```diff
@@ -0,0 +1,452 @@
+
+      context 'nil access_token' do
+        let(:payload) {
+          if authentication_service.is_default?
```
An `if` statement in a test is a code smell. Let's look through this together and see what we can break off into its own shared example and refactor using specific contexts for default and non-default auth services.

dmlond replied:
changed

dmlond replied:
refactored to use service_id in payload

dmlond replied:
refactored

dmann reviewed (changes requested):
I think it looks good overall. There are a couple things that should be added (like the Rails 5 compatibility bit) before merging this in to the develop branch. If time permits, it would be good for us to refactor users_api_spec and authentication_service_shared_examples.

dmlond committed:
added a test for a nil token passed to authentication_service.get_user_for_access_token(nil)
to ensure that all implementations raise an InvalidAccessTokenException

DukeAuthenticationService and OpenidAuthenticationService meet expectation

dmlond committed:
authentication_service model implementations must raise InvalidAccessTokenException when the token is nil

dmlond committed:
modified to use find_by(x: ...) instead of where(x: ...).[first|take]

dmlond committed:
ensure that auth_service:transfer_default updates to default on from_auth_service and to_auth_service
are in a transaction, with rollback if something bad happens

added RAILS 5 compatibility to the rake spec for silence stream

dmlond committed:
made the test for a nil token more generic and removed excess code

dmlond committed:
made all tests for authentication_service_endpoints based on the token use the explicit
authentication_service_id: token in the payload to remove excess code and switches based on
.is_default?

added let variable expectations to a before block to make the tests fail early if
these are net set by the calling context

dmlond committed:
created a utilities shared_example file to hold shared_context and shared_examples that are used
by applicable to any and all tests
this includes shared_context 'requires let variables' which takes an array of symbols
that any tests requires be 'let' in the calling context

refactored authentication_service_shared_examples to use the 'requires let variables' shared context
where appropriate
